Feature: Add IPFS integration

josemarinas opened the pull request:
## Description

Create an IPFS connection and add methods: 

- Pin
- FetchString
- FetchBytes

Task: [APP-317](https://aragonassociation.atlassian.net/browse/APP-317)

## Type of change


- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran all tests with success and extended them if necessary.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder of the package after the [TBD] title and before the latest version.
- [ ] I have tested my code on the test network.

marcvelmer reviewed (commented):
Before going deep in to the code:

1) The testing is not working in the CI, this should be reviewed.
2) Did you lint the code? It doesn't seem so.

josemarinas commented:
> Before going deep in to the code:
> 
>     1. The testing is not working in the CI, this should be reviewed.
> 
>     2. Did you lint the code? It doesn't seem so.
> 
>     3. Did you rebase the code with `develop`?

The code is not rebased yet, I fixed the lint but the build is failing due to the node version, the environment uses node 14 but the `ipfs-http-client` needs node 15 or higher, I updated the `package.json` to require at leat this version

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -166,4 +175,37 @@ export abstract class ClientCore implements IClientCore {
       }
     );
   }
+  public async pin(input: string | Uint8Array): Promise<string> {
```
```suggestion
  public pin(input: string | Uint8Array): Promise<string> {
```

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -166,4 +175,37 @@ export abstract class ClientCore implements IClientCore {
+      .add(input)
+      .then(res => {
+        return Promise.resolve(res.path);
+      })
```
```suggestion
      .then(res => res.path)
```

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -166,4 +175,37 @@ export abstract class ClientCore implements IClientCore {
+      .then(res => {
+        return Promise.resolve(res.path);
+      })
+      .catch(e => {
+        return Promise.reject(e);
+      });
```
```suggestion
      .catch(e => new Error("Could not pin data: " + (e?.message || "")));
```

josemarinas replied:
The function is not async, so I need to return a `Promise.reject`, this will force me to change the signature of the function

How about this?
```js
.catch(e => Promise.reject(new Error("Could not pin data: " + (e?.message || ""))));
```

brickpop replied:
The external function is not async, but whatever that happens inside the `then`/`catch` inner functions is handled accordingly. 
If you `throw new Error()` inside a `then` or `catch` clause is going to `reject` the external promise being returned

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -166,4 +175,37 @@ export abstract class ClientCore implements IClientCore {
+  }
+  public async fetchBytes(cid: string) {
+    if (!this.ipfs)
+      return Promise.reject(new Error("IPFS client is not initialized"));
```
```suggestion
      throw new Error("IPFS client is not initialized");
```

brickpop replied:
An async function will reject the base promise when an error is thrown

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -166,4 +175,37 @@ export abstract class ClientCore implements IClientCore {
+    if (!this.ipfs)
+      return Promise.reject(new Error("IPFS client is not initialized"));
+    try {
+      for await (const chunk of this.ipfs.cat(cid)) {
+        return Promise.resolve(chunk);
```
It looks like this is only going to return the first chunk, but we should definitely check on big files... and see if all chunks need to be serialized into one

josemarinas replied:
I tried with small files and all the chunks get serialized, but I'm not sure with big files

marcvelmer replied:
This part of the code is wrong for sure, we cannot return in the middle of a loop. Can you test with a big file to check what happens? Somehow we should store this into chunks and then create a `Buffer` from it before resolving.

josemarinas replied:
Tested with a big file and only returned the first chunk, fixed it in the last commit

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -166,4 +175,37 @@ export abstract class ClientCore implements IClientCore {
+        return Promise.resolve(chunk);
+      }
+    } catch (e) {
+      Promise.reject(e);
```
We're not returning anything here.
I would personally throw an Error of ourselves

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -166,4 +175,37 @@ export abstract class ClientCore implements IClientCore {
+    } catch (e) {
+      Promise.reject(e);
+    }
+    return;
```
```suggestion
```

josemarinas replied:
If i remove this, TS will give me this warning:
```
Not all code paths return a value.ts(7030)
```
I'm not sure why

josemarinas replied:
solved when fixed the big files problem

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -166,4 +175,37 @@ export abstract class ClientCore implements IClientCore {
+  }
+  public fetchString(cid: string): Promise<string> {
+    return this.fetchBytes(cid)
+      .then(bytes => {
+        return new TextDecoder().decode(bytes);
+      })
```
```suggestion
      .then(bytes => new TextDecoder().decode(bytes))
```

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -166,4 +175,37 @@ export abstract class ClientCore implements IClientCore {
+      .then(bytes => {
+        return new TextDecoder().decode(bytes);
+      })
+      .catch(e => {
+        return Promise.reject(e);
+      });
```
This code has no effect, I would either remove the catch block or throw a specific error.... but since nothing specific can fail here, I'd simply transparently relay the eventual error error

brickpop commented on javascript/modules/client/src/internal/interfaces/client-core.ts:
```diff
@@ -15,6 +15,7 @@ export interface IClientVocdoniCore {
 }
 export interface IClientIpfsCore {
   // Add here
+  pin(entry: string | Uint8Array): Promise<string>;
```
Aren't we missing the interaces for fetching as well?

brickpop commented on javascript/modules/client/src/internal/interfaces/context.ts:
```diff
@@ -1,5 +1,6 @@
 import { Signer } from "@ethersproject/abstract-signer";
 import { JsonRpcProvider, Networkish } from "@ethersproject/providers";
+import { Options, IPFSHTTPClient } from "ipfs-http-client";
```
```suggestion
import { Options as IPFSOptions, IPFSHTTPClient } from "ipfs-http-client";
```

brickpop replied:
I would rename the usage of `Options` into something like `IpfsOptions`. Otherwise the typing becomes ambiguous to someone consuming the library.

marcvelmer reviewed (commented):
@josemarinas 

We should have some kind of unit testing and it's pretty important at this point, because we will have issues with the usage of `TextDecoder` and `jest`. I already faced that problem.

So because it's a draft PR I would simply put any dummy test in order to tackle this problem and then we can redo it later.

FYI this helped me a lot:
```ts
const { TextEncoder, TextDecoder } = require('util');
global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;
```

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -244,4 +253,46 @@ export abstract class ClientCore implements IClientCore {
+      }
+      var mergedArray = new Uint8Array(totalArrayLength);
+      let lastIndex = 0;
+      chunks.forEach(chunk => {
+        mergedArray.set(chunk, lastIndex);
+        lastIndex += chunk.length;
+      });
```
It's a bit more efficient to avoid creating new function contexts for every loop iteration

```suggestion
      for (const chunk of chunks) {
        mergedArray.set(chunk, lastIndex);
        lastIndex += chunk.length;
      }
```

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -244,4 +253,46 @@ export abstract class ClientCore implements IClientCore {
+        chunks.push(chunk);
+        totalArrayLength += chunk.length;
+      }
+      var mergedArray = new Uint8Array(totalArrayLength);
```
```suggestion
      const mergedArray = new Uint8Array(totalArrayLength);
```

brickpop commented on javascript/modules/client/src/internal/interfaces/context.ts:
```diff
@@ -20,7 +16,8 @@ interface VocdoniContextParams {
 interface IpfsContextParams {
-  // ipfs: IpfsOptions;
+  ipfsOptions?: Options;
+  ipfsPeers?: string[];
```
Do we need it?

brickpop commented on javascript/modules/client/src/internal/interfaces/context.ts:
```diff
@@ -20,7 +16,8 @@ interface VocdoniContextParams {
 }
 interface IpfsContextParams {
-  // ipfs: IpfsOptions;
+  ipfsOptions?: Options;
```
Shouldn't this parameter be mandatory within the IPFS scope? Also, shouldn't we allow for receiving N >= 1 nodes info?

josemarinas replied:
If not specified, it will default to the lib default parameters that will try to connect to `localhost:5001`. We can overwrite these parameters ones, if not specified, set the default nodes to the aragon ones. About the having more than one node configuration makes sense, but I believe we should add a way to check the state of the IPFS node and In case that it is disconnected shift to the second one and the first selection between the available nodes should be random to avoid connecting always to the same, this does not make sense if Mathias plans to add a load balancer before the nodes.

brickpop commented:
@marcvelmer I'd be careful about the `TextEncoder` workaround. Having this on the testing is fine, but having this on the main source code is going to work on Node but fail on web browsers

brickpop commented on javascript/modules/client/src/context.ts:
```diff
@@ -1,12 +1,18 @@
+} from "ipfs-http-client";
+// const { TextEncoder, TextDecoder } = require("util");
+// global.TextEncoder = TextEncoder;
+// global.TextDecoder = TextDecoder;
```
Do we need to keep this now?

brickpop commented on javascript/modules/client/src/context.ts:
```diff
@@ -60,10 +65,9 @@ export class Context {
       throw new Error("No web3 endpoints defined");
     } else if (!contextParams.gasFeeEstimationFactor) {
       throw new Error("No gas fee reducer defined");
+    } else if (!contextParams.ipfsNodes) {
```
```suggestion
    } else if (!contextParams.ipfsNodes?.length) {
```

brickpop commented on javascript/modules/client/src/context.ts:
```diff
@@ -106,9 +110,9 @@ export class Context {
-    // if (contextParams.ipfs) {
-    //   this.state.ipfs = ipfsCreate(contextParams.ipfs);
-    // }
+    if (contextParams.ipfsNodes) {
```
```suggestion
    if (contextParams.ipfsNodes?.length) {
```

brickpop commented on javascript/modules/client/src/context.ts:
```diff
@@ -140,6 +144,17 @@ export class Context {
   }
 
+  useIpfsNodes(nodes: Options[]): IPFSHTTPClient[] {
+    if (nodes.length > 0) {
```
Should we accept a 0 length array? I'd throw an error

josemarinas replied:
I think so, the library have default values if not specified pointing to localhost, so we can keep this behavior, or force the user to always send a configuration, so this array will never be 0

brickpop replied:
I would not expect `localhost` to be available. If there is an empty array, I would throw that error

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -26,8 +27,14 @@ export abstract class ClientCore implements IClientCore {
   constructor(context: Context) {
+    if (context.ipfs) {
+      this._ipfs = context.ipfs;
+      this._ipfsIdx = Math.floor(Math.random() * context.ipfs.length);
```
If the array had 0 elements, we would have an index of `0`, which doesn't exist.
I'd throw an error if so

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -55,11 +62,18 @@ export abstract class ClientCore implements IClientCore {
   }
 
+  public shiftIpfsNode() {
+    if (!this._ipfs.length) throw new Error("No endpoints");
```
Might have not been initialized yet

```suggestion
    if (!this._ipfs?.length) throw new Error("No IPFS endpoints are available");
```

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -55,11 +62,18 @@ export abstract class ClientCore implements IClientCore {
 
+  public shiftIpfsNode() {
+    if (!this._ipfs.length) throw new Error("No endpoints");
+    else if (this._ipfs.length <= 1) {
```
```suggestion
    else if (this._ipfs?.length < 2) {
```

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -103,13 +117,29 @@ export abstract class ClientCore implements IClientCore {
+    return this._ipfs;
+  }
+
+  get ipfsIdx() {
```
Getters should have friendly names. How about `ipfsIndex`?
Indeed... do we really need to export this? Consumers care about getting a client, not about an index. See above

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -103,13 +117,29 @@ export abstract class ClientCore implements IClientCore {
   }
 
+  get ipfs() {
+    return this._ipfs;
```
I'd return a usable one, not just all.

```suggestion
    if (!this._ipfs[this._ipfsIdx]) throw new Error("No IPFS endpoints available");
    return this._ipfs[this._ipfsIdx];
```

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -244,4 +274,62 @@ export abstract class ClientCore implements IClientCore {
+      if (isOnline) break;
+      this.shiftIpfsNode();
+    }
+    if (!isOnline) throw new Error("IPFS nodes are not available");
```
```suggestion
    if (!isOnline) throw new Error("No IPFS nodes are available");
```

brickpop commented on javascript/modules/client/src/internal/client-core.ts:
```diff
@@ -244,4 +274,62 @@ export abstract class ClientCore implements IClientCore {
+      if (isOnline) break;
+      this.shiftIpfsNode();
+    }
+    if (!isOnline) throw new Error("IPFS nodes are not available");
```
```suggestion
    if (!isOnline) throw new Error("No IPFS nodes are available");
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/aragon/sdk?pullRequest=17) <br/>All committers have signed the CLA.
